massa-storage/src/lib.rs

```diff
@@ -1,66 +1,195 @@
+//! The clonable `Storage` module has thread-safe shared access to the stored objects.
+//!
+//! The `Storage` module also has lists of object references held by the current instance of `Storage`.
+//! When no instance of `Storage` claims a reference to a given object anymore, that object is automatically removed from storage.
 
 #![warn(missing_docs)]
 
 use massa_logging::massa_trace;
-use massa_models::prehash::{Map, Set};
+use massa_models::prehash::{Map, PreHashed, Set};
+use massa_models::wrapped::Id;
 use massa_models::{BlockId, OperationId, WrappedBlock, WrappedOperation};
-use parking_lot::RwLock;
-use std::collections::hash_map::Entry;
-use std::sync::Arc;
+use parking_lot::{RwLock, RwLockWriteGuard};
+use std::hash::Hash;
+use std::{collections::hash_map, sync::Arc};
 
-/// A storage of block, shared by various components.
-#[derive(Clone, Default)]
+/// A storage system for objects (blocks, operations...), shared by various components.
+#[derive(Default)]
 pub struct Storage {
+    /// global block storage
     blocks: Arc<RwLock<Map<BlockId, Arc<RwLock<WrappedBlock>>>>>,
+    /// global operation storage
     operations: Arc<RwLock<Map<OperationId, WrappedOperation>>>,
+
+    /// global block reference counter
+    block_owners: Arc<RwLock<Map<BlockId, usize>>>,
+    /// global operation reference counter
+    operation_owners: Arc<RwLock<Map<OperationId, usize>>>,
+
+    /// locally used block references
+    local_used_blocks: Set<BlockId>,
+    /// locally used block references
+    local_used_ops: Set<OperationId>,
+}
+
+impl Clone for Storage {
+    /// Clones the Storage instance
+    /// Note that the local references are reset for the new instance.
+    fn clone(&self) -> Self {
+        Self {
+            blocks: self.blocks.clone(),
+            operations: self.operations.clone(),
+            operation_owners: self.operation_owners.clone(),
+            block_owners: self.block_owners.clone(),
+
+            // local reference lists are not cloned
+            local_used_ops: Default::default(),
+            local_used_blocks: Default::default(),
+        }
+    }
 }
 
 impl Storage {
-    /// Store a block, along with it's serialized representation.
-    pub fn store_block(&self, block: WrappedBlock) {
-        massa_trace!("storage.storage.store_block", { "block_id": block.id });
-        let mut blocks = self.blocks.write();
-        match blocks.entry(block.id) {
-            Entry::Occupied(_) => {}
-            Entry::Vacant(entry) => {
-                let to_store = Arc::new(RwLock::new(block));
-                entry.insert(to_store);
+    /// internal helper to locally claim a reference to an object
+    fn internal_claim_refs<IdT: Id + PartialEq + Eq + Hash + PreHashed + Copy>(
+        ids: &[IdT],
+        mut owners: RwLockWriteGuard<Map<IdT, usize>>,
+        local_used_ids: &mut Set<IdT>,
+    ) {
+        for &id in ids {
+            if local_used_ids.insert(id) {
```

Comment:
I don't understand this if. It should be "if not" instead ? Because here I understand that we increase only if it's the second time we locally claim the id. @damip

Reply:
basically, if we have already claimed this ID, we don't increment counters on a second claim from the same module

Reply:
Ah yes it returns true if there is no value before it's the contrary of insert with hashmap... My bad.